### PR TITLE
Add logging controller

### PR DESCRIPTION
**Pull request by dragana8**
Establishes a Logging Controller that can be used throughout MetaMask products to store data meant primarily for debugging purposes and for support staff to provide more details for tickets as well as to assist customers in troubleshooting MetaMask. the primary use case for this implementation is signature request logs, however a generic log type has been added as well. It is recommended to add a log type for each special case and to only use GenericLog type for general messages.

There are only 3 methods on this controller, 1 of which is private:
#generateId - handles random id generation and prevents collisions in rare cases. 
clear - clears all logs. This method should likely only be called by things like the feature to wipe state. 
add - Adds a log 

The add method is attached to the messenger to allow other controllers with access to the messenging system to add logs without having a reference to the logging controller.

### CODE REVIEW

File: packages/logging-controller/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Changelog
```
Comment (mcmire) [inline packages/logging-controller/CHANGELOG.md:1]: I think we will want to start this off with an empty changelog for now. It's customary for us to update the changelog when we issue a new release rather than have it already ready to go prior to the new release.

File: packages/logging-controller/LICENCE [hunk 2]
```diff
@@ -0,0 +1,20 @@
+MIT License
+
+Copyright (c) 2018 MetaMask
```
Comment (mcmire) [inline packages/logging-controller/LICENCE:3]: Update this year?

File: packages/logging-controller/src/LoggingController.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import {
+  BaseController,
```
Comment (mcmire) [inline packages/logging-controller/src/LoggingController.ts:2]: Be advised that there are two versions of BaseController in this repo, v1 and v2. If you just see a controller import `BaseController`, then that's v1, but we want to use v2 for all new controllers going forward. You can see an example of a controller that uses v2 here: https://github.com/MetaMask/core/blob/main/packages/ens-controller/src/EnsController.ts

File: packages/logging-controller/src/LoggingController.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  BaseConfig,
+  BaseState,
+} from '@metamask/base-controller';
+import { v1 as random } from 'uuid';
+
+export enum LogType {
+  eth_sign_proposed = 'eth_sign_proposed',
```
Comment (mcmire) [inline packages/logging-controller/src/LoggingController.ts:9]: We don't have this ESLint rule enabled yet, but as a rule we use PascalCase for enum property names. So to fit idiomatically with other enums, we ought to use:

``` typescript
export enum LogType {
  EthSignProposed = 'eth_sign_proposed',
  EthSignSigned = 'eth_sign_signed',
  // ...
}
```

File: packages/logging-controller/src/LoggingController.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * @type Log
+ *
+ * Log representation
+ * @property id - Generated UUID associated with this log
+ * @property timestamp - Timestamp associated with this log
+ * @property type - Type of signature request
+ * @property data - Additional data about signature request
+ */
```
Comment (mcmire) [inline packages/logging-controller/src/LoggingController.ts:34]: Note: `@type` is used to denote the type of a variable or fragment of code, but the proper tag for type _definitions_ is `@typedef`. That said, we use TypeDoc to generate docs and it doesn't really honor either `@type` or `@typedef`, as we can just use types. The format here is the same as for a class, function or method:

```
Description of thing

Tag1
Tag2
Tag3
...
```

So this can be:

``` suggestion
/**
 * Log representation
 * 
 * @property id - Generated UUID associated with this log
 * @property timestamp - Timestamp associated with this log
 * @property type - Type of signature request
 * @property data - Additional data about signature request
 */

File: packages/logging-controller/package.json [hunk 6]
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@metamask/logging-controller",
+  "version": "1.0.0",
```
Comment (mcmire) [inline packages/logging-controller/package.json:3]: In order to publish the first version of this package, we will want to start this off at 0.0.0 and then bump this to 1.0.0 as part of a release PR. Changing the version of a package inside of a release PR is how our tooling knows that a new release ought to be issued for that package — if this is already 1.0.0 when that PR is created and doesn't get bumped, then we may run into problems.

**Review by mcmire — COMMENTED** [review]
I know this is in draft but I couldn't help notice this! I think there are a few things we will want to take care of before merging this.

File: packages/logging-controller/README.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+`yarn add @metamask/logging-controller`
+
+or
+
+`npm install @metamask/logging-controller`
+
+## Contributing
+
+This package is part of a monorepo. Instructions for contributing can be found in the [monorepo README](https://github.com/MetaMask/core#readme).
```
[the remark this answers is not in the document]
Comment (mcmire) [reply packages/logging-controller/README.md:15]: We didn't copy over any usage instructions when creating the monorepo — that was planned for later — so I can understand why this was omitted. But yes I agree that we should do that. You can do this by adding a new `## Usage` section below the `## Installation` section. Usually this just consists of an example (such as for [`@metamask/providers`](https://github.com/MetaMask/providers#usage)) but could also describe certain ways of using the package if that's necessary (such as for [`action-npm-publish`](https://github.com/MetaMask/action-npm-publish#usage)).

**Comment by mcmire** [comment]
Make sure to run `yarn generate-dependency-graph` before merging this to update the graph at the bottom of the README :)

File: packages/logging-controller/LICENCE (hunk 2 above)
Comment (brad-decker) [reply packages/logging-controller/LICENCE]: Done

File: packages/logging-controller/src/LoggingController.ts (hunk 3 above)
Comment (brad-decker) [reply packages/logging-controller/src/LoggingController.ts]: Refactored this whole file to use BaseControllerV2 and a slightly different overall architecture.

File: packages/logging-controller/src/LoggingController.ts (hunk 4 above)
Comment (brad-decker) [reply packages/logging-controller/src/LoggingController.ts]: I updated these types and data structure quite a bit

File: packages/logging-controller/src/LoggingController.ts (hunk 5 above)
Comment (brad-decker) [reply packages/logging-controller/src/LoggingController.ts]: Most of the jsdocs should be updated. WOuld appreciate a double check

File: packages/logging-controller/package.json (hunk 6 above)
Comment (brad-decker) [reply packages/logging-controller/package.json]: Oh oops I did 0.0.1 i'll drop it to 0

**Comment by brad-decker** [comment]
```
Error: not found: dot
    at getNotFoundError (/Users/braddecker/Documents/dev/consensys/core/node_modules/which/lib/index.js:16:17)
    at which (/Users/braddecker/Documents/dev/consensys/core/node_modules/which/lib/index.js:76:9) {
  code: 'ENOENT'
}```

@mcmire I get this trying to run that `yarn generate-dependency-graph' script

**Comment by brad-decker** [comment]
Ah i see I need graphviz

**Comment by brad-decker** [comment]
@mcmire could you give this another pass when you get a moment? I need to write tests for this still but I want to make sure my reasoning is good and that is generally going in the right direction.

**Comment by brad-decker** [comment]
@mcmire could you give this another pass when you get a moment? I need to write tests for this still but I want to make sure my reasoning is good and that is generally going in the right direction.

File: packages/logging-controller/CHANGELOG.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+## [Unreleased]
```
Comment (mcmire) [inline packages/logging-controller/CHANGELOG.md:1]: I think this needs to include the header and the preamble in order for `yarn changelog:validate` to pass:

```
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

[Unreleased]: https://github.com/MetaMask/core/
```

File: packages/logging-controller/src/LoggingController.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+   * Removes all log entries.
+   */
+  clear() {
+    this.update((state) => {
+      state.logs = {};
+    });
+  }
+}
+
+export default LoggingController;
```
Comment (mcmire) [inline packages/logging-controller/src/LoggingController.ts:180]: Do we need to export this as the default, since we're already exported it as a named export?

**Review by mcmire — COMMENTED** [review]
@brad-decker Looks good! Just a couple of things I noticed.

File: packages/logging-controller/README.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
+# `@metamask/logging-controller`
+
+Stores logs about signature request.
```
Comment (mcmire) [inline packages/logging-controller/README.md:3]: It looks like the description in `package.json` says "Manages a list of logs for signature requests". Maybe this should be changed too? It seems like there'd be no reason to make the description between the README and `package.json` different.

``` suggestion
Manages a list of logs for signature requests.
```

File: packages/logging-controller/src/LoggingController.ts [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+   * match that id.
+   *
+   * @returns unique id
+   */
+  generateId(): string {
+    const id = random();
+    if (this.state.logs[id]) {
+      return this.generateId();
+    }
+    return id;
```
Comment (mcmire) [inline packages/logging-controller/src/LoggingController.ts:150]: Considering that no JavaScript engine has tail-call recursion, instead of making this recursive, what do we think about doing this in a loop? Also, this tests that `this.state.logs[id]` is truthy, which isn't quite accurate; it seems that we want to check that an entry is present under that ID. So what about:

``` suggestion
    let id = random();
    while (id in this.state.logs) {
      id = random();
    }
    return id;
```

Also, does this method need to be public? It seems like it'd only be used by `addLog`.

File: packages/logging-controller/src/LoggingController.ts [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+    this.update((state) => {
+      state.logs[newLog.id] = newLog;
+    });
+  }
+
+  /**
+   * Removes all log entries.
+   */
+  clear() {
```
Comment (mcmire) [inline packages/logging-controller/src/LoggingController.ts:173]: Maybe this should be called `clearLogs` for consistency with `addLog`? Or else, call `addLog` just `add`.

File: packages/logging-controller/README.md (hunk 10 above)
Comment (brad-decker) [reply packages/logging-controller/README.md:3]: Actually we should change the package.json to match this, because its more than just signature requests that can store logs.

File: packages/logging-controller/src/LoggingController.ts (hunk 11 above)
Comment (brad-decker) [reply packages/logging-controller/src/LoggingController.ts:150]: Can you ELI5 me the tail-call recursion part of this optimization? I'm not countering it at all just wanna understand. In the meantime I'm accepting your suggestion.

File: packages/logging-controller/src/LoggingController.ts (hunk 12 above)
Comment (brad-decker) [reply packages/logging-controller/src/LoggingController.ts:173]: I'll change addLog to add, i think its kind of self explanatory what we are adding.

File: packages/logging-controller/src/LoggingController.ts (hunk 9 above)
Comment (mcmire) [reply packages/logging-controller/src/LoggingController.ts:180]: Update: It seems that we do this quite a lot in our controllers, including EnsController, so I can see where this came from. That said, I consider this an anti-pattern, as there's no real need for it (we don't expect developers to import individual files from packages, even though it's technically possible).

File: packages/logging-controller/README.md (hunk 10 above)
Comment (brad-decker) [reply packages/logging-controller/README.md:3]: Oh i see. i hadnt pushed changes I made to the readme yet 😓

File: packages/logging-controller/src/LoggingController.ts (hunk 11 above)
Comment (mcmire) [reply packages/logging-controller/src/LoggingController.ts]: Every time a function is called it gets added to the stack, but the stack has a max size, so if a recursive function is called too many times it will cause a stack overflow error. In some languages, though, if a recursive call occurs as the very last step of the function, and the function returns the result of the call, the language engine can detect this and prevent the stack from growing. This is called tail-call optimization (or sometimes "proper tail calls", which is a more general idea that doesn't apply to just recursive calls). I misspoke when I said that no JavaScript engine supports this: it seems that Safari does, but it's alone — neither Chrome nor Firefox do, and don't plan on adding support any time soon.

Now, I don't know if this code is likely to produce a stack overflow error. UUIDs are supposed to be unique, so the chances are probably very low, and using recursion is probably okay. It's something to be aware of though, and using a loop instead of recursion prevents that issue from coming up.

**Review by pedronfigueiredo — APPROVED** [review]

File: packages/logging-controller/src/LoggingController.ts (hunk 11 above)
Comment (brad-decker) [reply packages/logging-controller/src/LoggingController.ts]: Thanks @mcmire!

**Review by DDDDDanica — DISMISSED** [review]

**Review by danjm — CHANGES_REQUESTED** [review]
Screenshot from 2023-05-31 14-54-25

**Review by danjm — CHANGES_REQUESTED** [review]
Screenshot from 2023-05-31 14-54-25

**Review by danjm — APPROVED** [review]
Nice work!

**Review by pedronfigueiredo — DISMISSED** [review]

**Review by pedronfigueiredo — APPROVED** [review]

**Review by mcmire — APPROVED** [review]
LGTM!
